src/nvim/eval/typval.c

```diff
@@ -0,0 +1,2556 @@
+///
+/// @param[in]  tv  Value to check.
+///
+/// @return True if value is locked, false otherwise.
+bool tv_islocked(const typval_T *const tv)
+  FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
+{
+  return ((tv->v_lock == VAR_LOCKED)
+          || (tv->v_type == VAR_LIST
+              && tv->vval.v_list != NULL
+              && (tv->vval.v_list->lv_lock == VAR_LOCKED))
+          || (tv->v_type == VAR_DICT
+              && tv->vval.v_dict != NULL
+              && (tv->vval.v_dict->dv_lock == VAR_LOCKED)));
+}
+
+/// Return true if typval is locked
+///
+/// Also gives an error message when typval is locked.
+///
+/// @param[in]  lock  Lock status.
+/// @param[in]  name  Variable name, used in the error message.
+/// @param[in]  name_len  Variable name length.
+///
+/// @return true if variable is locked, false otherwise.
+bool tv_check_lock(const VarLockStatus lock, const char *const name,
+                   const size_t name_len)
+  FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  const char *error_message = NULL;
+  switch (lock) {
+    case VAR_UNLOCKED: {
+      return false;
+    }
+    case VAR_LOCKED: {
+      error_message = N_("E741: Value is locked: %.*s");
+      break;
+    }
+    case VAR_FIXED: {
+      error_message = N_("E742: Cannot change value of %.*s");
+      break;
+    }
+  }
+  assert(error_message != NULL);
+
+  const char *const unknown_name = _("Unknown");
+
+  emsgf(_(error_message), (name != NULL ? name_len : strlen(unknown_name)),
+        (name != NULL ? name : unknown_name));
+
+  return true;
+}
+
+//{{{2 Comparison
+
+static int tv_equal_recurse_limit;
+
+/// Compare two VimL values
+///
+/// Like "==", but strings and numbers are different, as well as floats and
+/// numbers.
+///
+/// @warning Too nested structures may be considered equal even if they are not.
+///
+/// @param[in]  tv1  First value to compare.
+/// @param[in]  tv2  Second value to compare.
+/// @param[in]  ic  True if case is to be ignored.
+/// @param[in]  recursive  True when used recursively.
+///
+/// @return true if values are equal.
+bool tv_equal(typval_T *const tv1, typval_T *const tv2, const bool ic,
+              const bool recursive)
+  FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
+{
+  // TODO(ZyX-I): Make this not recursive
+  static int recursive_cnt = 0;  // Catch recursive loops.
+
+  if (!(tv_is_func(*tv1) && tv_is_func(*tv2)) && tv1->v_type != tv2->v_type) {
+    return false;
+  }
+
+  // Catch lists and dicts that have an endless loop by limiting
+  // recursiveness to a limit.  We guess they are equal then.
+  // A fixed limit has the problem of still taking an awful long time.
+  // Reduce the limit every time running into it. That should work fine for
+  // deeply linked structures that are not recursively linked and catch
+  // recursiveness quickly.
+  if (!recursive) {
+    tv_equal_recurse_limit = 1000;
+  }
+  if (recursive_cnt >= tv_equal_recurse_limit) {
+    tv_equal_recurse_limit--;
```

Comment:
The result of this function depends on the first call to it being made with `recursive == false` to set `tv_equal_recurse_limit == 1000`, otherwise unequal tvs could compare to true.
Could also happen later on if `tv_equal_recurse_limit` decreases to `0` and calls to this function are made with `recursive == true`.